lib/byron/bench/Restore.hs

```diff
@@ -0,0 +1,468 @@
+        let convert = fromByronBlock (getGenesisBlockHash bp) (getEpochLength bp)
+        let nw = convert <$> nw'
+        waitForNodeSync @n nw logQuiet
+    sayErr . fmt $ "Completed sync of "+|networkDiscriminantVal @n|+" up to "+||sl||+""
+
+-- | Regularly poll the wallet to monitor it's syncing progress. Block until the
+-- wallet reaches 100%.
+waitForWalletSync
+    :: forall (n :: NetworkDiscriminant) s t k. (KnownNetwork n)
+    => WalletLayer s t k
+    -> WalletId
+    -> IO ()
+waitForWalletSync walletLayer wid = do
+    (w, _, _) <- unsafeRunExceptT $ W.readWallet walletLayer wid
+    let tol = mkSyncTolerance 3600
+    let bp = blockchainParameters @n
+    prog <- syncProgressRelativeToTime
+                tol
+                (slotParams bp)
+                (currentTip w)
+                <$> getCurrentTime
+    case prog of
+        Ready -> return ()
+        NotResponding -> do
+            threadDelay 1000000
+            waitForWalletSync @n walletLayer wid
+        Syncing (Quantity p) -> do
+            sayErr . fmt $ "[INFO] restoring: "+|p|+""
+            threadDelay 1000000
+            waitForWalletSync @n walletLayer wid
+
+-- | Poll the network tip until it reaches the slot corresponding to the current
+-- time.
+waitForNodeSync
+    :: forall (n :: NetworkDiscriminant). (KnownNetwork n)
+    => NetworkLayer IO (IO Byron) Block
+    -> (SlotId -> SlotId -> IO ())
+    -> IO SlotId
+waitForNodeSync nw logSlot = loop 10
+  where
+    loop :: Int -> IO SlotId
+    loop retries = runExceptT (currentNodeTip nw) >>= \case
+        Right (BlockHeader tipBlockSlot _ _ _) -> do
+            currentSlot <- getCurrentSlot @n
+            logSlot tipBlockSlot currentSlot
+            if tipBlockSlot < currentSlot
+                then do
+                    -- 2 seconds poll interval
+                    threadDelay 2000000
+                    loop retries
+                else
+                    pure tipBlockSlot
+        Left e | retries > 0 -> do
+                     sayErr "Fetching tip failed, retrying shortly..."
+                     threadDelay 15000000
+                     loop (retries - 1)
+               | otherwise -> throwIO e
+
+-- | Calculate the current slot, because the network layer doesn't know it.
+getCurrentSlot
+    :: forall (n :: NetworkDiscriminant). (KnownNetwork n)
+    => IO SlotId
+getCurrentSlot = calcSlot <$> startTime <*> getPOSIXTime
+  where
+    calcSlot :: POSIXTime -> POSIXTime -> SlotId
+    calcSlot start now = SlotId (EpochNo ep) (SlotNo idx)
+      where
+        d = now - start
+        slotDur = 20
+        epochDur = slotDur * 21600
+        ep = floor (d / epochDur)
+        idx = floor ((d - (fromIntegral ep) * epochDur) / slotDur)
+
+    startTime :: MonadFail m => m POSIXTime
+    startTime = pure
+        $ (\(StartTime t) -> utcTimeToPOSIXSeconds t)
+        $ getGenesisBlockDate
+        $ blockchainParameters @n
```

Comment:
I assume this function was taken from the old http-bridge implementation. We know have:

```
slotAt :: SlotParameters -> UTCTime -> Maybe SlotId
```

in `Cardano.Wallet.Primitive.Types` which can get us exactly the same result, what about using this instead of duplicating the logic :pray:  ?

Reply:
ebf7978769b6876b7c31db914e41777780a058cc